versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/MergeBase.java

```diff
@@ -0,0 +1,325 @@
+import static org.projectnessie.versioned.storage.common.logic.ShallowCommit.BOTH_COMMITS;
+import static org.projectnessie.versioned.storage.common.logic.ShallowCommit.CANDIDATE;
+import static org.projectnessie.versioned.storage.common.persist.ObjId.EMPTY_OBJ_ID;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.agrona.collections.Object2ObjectHashMap;
+import org.immutables.value.Value;
+import org.projectnessie.versioned.storage.common.objtypes.CommitObj;
+import org.projectnessie.versioned.storage.common.persist.ObjId;
+
+/**
+ * Identifies the nearest commit that shall be used as the base commit for a merge of the given
+ * {@link #fromCommitId() from-commit} onto the given {@link #targetCommitId() target commit}.
+ *
+ * <p>This class also supports finding the base commit for N-way merges, although N-way merges are
+ * not implemented for Nessie (yet).
+ */
+@Value.Immutable
+public abstract class MergeBase {
+  public abstract Function<ObjId, CommitObj> loadCommit();
+
+  public abstract ObjId targetCommitId();
+
+  public abstract ObjId fromCommitId();
+
+  /**
+   * Whether merge-parents shall be respected, defaults to {@code true}. Setting this to {@code
+   * false} changes the behavior to return the <em>common ancestor</em> instead of the nearest
+   * merge-base.
+   */
+  @Value.Default
+  public boolean respectMergeParents() {
+    return true;
+  }
+
+  public static ImmutableMergeBase.Builder builder() {
+    return ImmutableMergeBase.builder();
+  }
+
+  @Value.NonAttribute
+  public ObjId identifyMergeBase() {
+    List<ShallowCommit> mergeBases = identifyAllMergeBases();
+    if (mergeBases == null || mergeBases.isEmpty()) {
+      throw noCommonAncestor();
+    }
+    return mergeBases.get(0).id();
+  }
+
+  private List<ShallowCommit> identifyAllMergeBases() {
+    ShallowCommit targetCommit = shallowCommit(targetCommitId());
+    if (targetCommit == null) {
+      // This is rather a UX hack, to raise a "not found" exception if any of the from-commit-IDs
+      // does not exist
+      shallowCommit(fromCommitId());
+      return null;
+    }
+
+    ShallowCommit fromCommit = shallowCommit(fromCommitId());
+    if (fromCommit == null) {
+      return singletonList(targetCommit);
+    }
+
+    return findMergeBases(fromCommit, targetCommit);
+  }
+
+  private List<ShallowCommit> findMergeBases(ShallowCommit commitA, ShallowCommit commitB) {
+    if (commitB.id().equals(commitA.id())) {
+      return newArrayList(commitA);
+    }
+
+    List<ShallowCommit> reachableCommits = flagReachableCommits(commitA, commitB);
```

Comment:
returning `reachableCommits` at this point does not break `TestMergeBase`. I tend to think this list already contains best merge base as the first element, but if I missed something perhaps we need another test case... WDYT?

Reply:
Seems we can at least short-cut when there's only one reachable commit.

Reply:
The short-cut is fine, but I'm still not sure that `removeRedundant` is required anymore.

`flagReachableCommits` already returns results in sequence order (greatest first) because of the new deque sort order, which basically means the merge bases closer to the target/source HEADs will be closer to the start of the `reachableCommits` list.

The only uncertainty is when the history graph is symmetrical and we have more than one merge base with the same sequence number (for example the case in `TestMergeBase.afterCrossMerge()`). I think it might be preferable to resolve that by including commit time into the deque comparator (and commit ID to ensure total order). Then, the output of `flagReachableCommits` will be deterministic with the highest sequence number at the first element. WDYT?

Alternatively, we could prefer merge bases that are direct descendants of the "target" commit (add a new flag to mark them).

Reply:
Not sure whether it's worth to change it. The logic can still return less results - just not sure how to construct a test case for that.

Reply:
> The logic can still return less results [...]

Well, I'm not sure this is actually possible... or rather, it feels like we can adjust the comparator in `flagReachableCommits` to make the `removeRedundant` unnecessary and greatly simplify the code.

Reply:
Do you insist?